// dotenv-vault/dist/commands/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const build_service_1 = require("../services/build-service");
class Build extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Build);
        const dotenvMe = flags.dotenvMe;
        const yes = flags.yes;
        await new build_service_1.BuildService({ cmd: this, dotenvMe: dotenvMe, yes: yes }).run();
    }
}
exports.default = Build;
Build.description = 'Build .env.vault';
Build.examples = [
    '<%= config.bin %> <%= command.id %>',
];
Build.flags = {
    dotenvMe: core_1.Flags.string({
        char: 'm',
        description: 'Pass .env.me (DOTENV_ME) credential directly (rather than reading from .env.me file)',
        hidden: false,
        multiple: false,
        env: 'DOTENV_ME',
        required: false,
    }),
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
