// dotenv-vault/dist/commands/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const login_service_1 = require("../services/login-service");
class Login extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(Login);
        const dotenvMe = args.DOTENV_ME;
        const yes = flags.yes;
        await new login_service_1.LoginService({ cmd: this, dotenvMe: dotenvMe, yes: yes }).run();
    }
}
exports.default = Login;
Login.description = 'Log in';
Login.examples = [
    '<%= config.bin %> <%= command.id %>',
];
Login.args = [
    {
        name: 'DOTENV_ME',
        required: false,
        description: 'Set .env.me credential. Defaults to generated value.',
        hidden: false,
    },
];
Login.flags = {
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
