// dotenv-vault/dist/commands/new.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const new_service_1 = require("../services/new-service");
class New extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(New);
        const dotenvVault = args.DOTENV_VAULT;
        const yes = flags.yes;
        await new new_service_1.NewService({ cmd: this, dotenvVault: dotenvVault, yes: yes }).run();
    }
}
exports.default = New;
New.description = 'Create your project';
New.examples = [
    '<%= config.bin %> <%= command.id %>',
];
New.args = [
    {
        name: 'DOTENV_VAULT',
        required: false,
        description: 'Set .env.vault identifier. Defaults to generated value.',
        hidden: false,
    },
];
New.flags = {
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
