// dotenv-vault/dist/commands/open.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const open_service_1 = require("../services/open-service");
class Open extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(Open);
        const environment = args.environment;
        const yes = flags.yes;
        await new open_service_1.OpenService({ cmd: this, environment: environment, yes: yes }).run();
    }
}
exports.default = Open;
Open.description = 'Open project page';
Open.examples = [
    '<%= config.bin %> <%= command.id %>',
];
Open.args = [
    {
        name: 'environment',
        required: false,
        description: 'Set environment to open to. Defaults to development',
        hidden: false,
    },
];
Open.flags = {
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
