// dotenv-vault/dist/commands/push.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const push_service_1 = require("../services/push-service");
class Push extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(Push);
        const environment = args.environment;
        const filename = args.filename;
        const dotenvMe = flags.dotenvMe;
        const yes = flags.yes;
        await new push_service_1.PushService({ cmd: this, environment: environment, filename: filename, dotenvMe: dotenvMe, yes: yes }).run();
    }
}
exports.default = Push;
Push.description = 'Push .env securely';
Push.examples = [
    '<%= config.bin %> <%= command.id %>',
];
Push.args = [
    {
        name: 'environment',
        required: false,
        description: 'Set environment to push from. Defaults to development',
        hidden: false,
    },
    {
        name: 'filename',
        required: false,
        description: 'Set input filename. Defaults to .env for development and .env.{environment} for other environments',
        hidden: false,
    },
];
Push.flags = {
    dotenvMe: core_1.Flags.string({
        char: 'm',
        description: 'Pass .env.me (DOTENV_ME) credential directly (rather than reading from .env.me file)',
        hidden: false,
        multiple: false,
        env: 'DOTENV_ME',
        required: false,
    }),
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
