// dotenv-vault/dist/commands/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const versions_service_1 = require("../services/versions-service");
class Versions extends core_1.Command {
    async run() {
        const { args, flags } = await this.parse(Versions);
        const environment = args.environment;
        const dotenvMe = flags.dotenvMe;
        const yes = flags.yes;
        await new versions_service_1.VersionsService({ cmd: this, environment: environment, dotenvMe: dotenvMe, yes: yes }).run();
    }
}
exports.default = Versions;
Versions.description = 'List version history';
Versions.examples = [
    '<%= config.bin %> <%= command.id %>',
];
Versions.args = [
    {
        name: 'environment',
        required: false,
        description: 'Set environment to check versions against. Defaults to development',
        hidden: false,
    },
];
Versions.flags = {
    dotenvMe: core_1.Flags.string({
        char: 'm',
        description: 'Pass .env.me (DOTENV_ME) credential directly (rather than reading from .env.me file)',
        hidden: false,
        multiple: false,
        env: 'DOTENV_ME',
        required: false,
    }),
    yes: core_1.Flags.boolean({
        char: 'y',
        description: 'Automatic yes to prompts. Assume yes to all prompts and run non-interactively.',
        hidden: false,
        required: false,
        default: false,
    }),
};
